resolwe/flow/models.py

```diff
@@ -479,6 +479,35 @@ def add_dependency(value):
                 for data in value:
                     add_dependency(data)
 
+    def validate_json_fields(self):
+        try:
+            if self.pk is None:  # on create
+                validate_schema(self.input, self.process.input_schema)  # pylint: disable=no-member
+
+            if self.descriptor_schema:
+                validate_schema(self.descriptor, self.descriptor_schema.schema)  # pylint: disable=no-member
+            elif self.descriptor and self.descriptor != {}:
+                raise ValueError("`descriptor_schema` must be defined if `descriptor` is given")
+
+            path_prefix = os.path.join(settings.FLOW_EXECUTOR['DATA_DIR'], str(self.pk))
+            output_schema = self.process.output_schema  # pylint: disable=no-member
+            if self.status == Data.STATUS_DONE:
+                validate_schema(self.output, output_schema, path_prefix=path_prefix)
+            else:
+                validate_schema(self.output, output_schema, path_prefix=path_prefix,
+                                test_required=False)
+        except ValidationError as ex:
+            # if object doesn't have ``pk`` assigned and doesn't validate, it
+            # won't be stored in database, so we don't have to do anything
+            if self.pk:
+                # current object is already corupted, so we have to get fresh
+                # copy (last that has passed validation)
+                data = Data.objects.get(pk=self.pk)
```

Comment:
@mstajdohar we discussed this before and agreed that this would be the best place to handle this error, as otherwise it should be handled on (too) many places.
I don't see a problem updating old object, because only "metadata" is updated (actual data, i.e. input and output, aren't touched).
In your example, `ValidationError` would be (re)raised, and the nice thing about this is that Data object would stay consistent even when updating it through shell ( = `status` and `error` would be updated automatically as you probably wouldn't handle it manually).

What do you (and @kostko) think?